File: project/samples/python/python-TLS/linux-x86-64/HelloThread.py
#-*- coding:utf-8 -*-

import threading

import json

from IoTMakersDeviceClient import IoTMakersDeviceClient

imClient = IoTMakersDeviceClient();

thread_run_flag = False

##########################################
#   리소스 설정(제어) 요청 처리 핸들러
##########################################
def OnResourceSetRequestHandler(pktBody, dev_id, resource_id, properties_in_jstr): 
    print ("OnResourceSetRequestHandler()-->", dev_id, resource_id, properties_in_jstr)

    # YOUR CONTROL CODE HERE
    properties = json.loads(properties_in_jstr);

    print ( json.dumps(properties, indent=4));

    #IM_RESP_CODE_2004_Changed
    return 2004;


##########################################
#   특정 리소스(resource_id) 조회 요청 처리 핸들러
##########################################
def OnResourceRetrieveOneRequestHandler(pktBody, dev_id, resource_id): 
    print ("OnResourceSetRequestHandler()-->", dev_id, resource_id)

    # YOUR CONTROL CODE HERE
    properties = { 
            "ectestString": "hello",
            "ectestDouble": 99.55,
            "ectestInteger": 100,
            "ectestBoolean": True
        };

    rc = imClient.ImResourceRetrieveSetResource(pktBody, resource_id, json.dumps(properties))
    if rc < 0 :
        print("fail ImResourceRetrieveSetResource()");
        ## Internal Error
        return 5000;
    
    #IM_RESP_CODE_2000_OK
    return 2000;


##########################################
#   본 디바이스(dev_id)의 모든 리소스 조회 요청 처리 핸들러
##########################################
def OnResourceRetrieveAllRequestHandler(pktBody, dev_id): 
    print ("OnResourceSetRequestHandler()-->", dev_id)

    # YOUR CONTROL CODE HERE

    # 1. append FIRST RESOURCE
    resource_id = "testAllTypeUri";

    properties = { 
            "ectestString": "hello",
            "ectestDouble": 99.55
        };

    rc = imClient.ImResourceRetrieveAppendResource(pktBody, resource_id, json.dumps(properties))
    if rc < 0 :
        print("fail ImResourceRetrieveAppendResource()");
        ## Internal Error
        return 5000;


    # 2. append SECOND RESOURCE
    resource_id = "testAllTypeUri";

    properties = { 
            "ectestInteger": 100,
            "ectestBoolean": True
        };

    rc = imClient.ImResourceRetrieveAppendResource(pktBody, resource_id, json.dumps(properties))
    if rc < 0 :
        print("fail ImResourceRetrieveAppendResource()");
        ## Internal Error
        return 5000;

    #IM_RESP_CODE_2000_OK
    return 2000;



##########################################
#   리소스 값 전송 예제
##########################################
def example_resource_send_notification(): 

    resource_id = "testAllTypeUri";

    properties = { 
            "ectestString": "hello",
            "ectestDouble": 99.55,
            "ectestInteger": 100,
            "ectestBoolean": True
        };


    print(json.dumps(properties));
    
    # 1. 전송 패킷 준비
    rc = imClient.ImResourceNotificationInit()
    if rc < 0 :
        print("fail ImResourceNotificationInit()");
        return -1;


    # 2. 패킷에 리소스 추가; append FIRST RESOURCE
    rc = imClient.ImResourceNotificationAppendResource(resource_id, json.dumps(properties));
    if rc < 0 :
        print("fail ImResourceNotificationAppendResource()");
        return -1;


    # 3. 패킷 전송
    rc = imClient.ImResourceNotificationSend();
    if rc < 0 :
        print("fail ImResourceNotificationSend()");
        return -1;

    
    print("OK, SENT: ", json.dumps(properties));


    return 0;





##########################################
# 수신 처리용 쓰레드
##########################################
def _thread_poll_sample(n):
    rc = 0
    global thread_run_flag

    while thread_run_flag == True:
        rc = imClient.ImPoll();
        if rc < 0 :
            print("FAIL ImPoll()...");
            thread_run_flag = False
            break;

        #print("_thread_poll_sample()...");
        imClient.ImMSleep(1000);

    print("_thread_poll_sample()..., DONE.");
 


##########################################
# 리소스 Notofication 전송 쓰레드
##########################################
def _thread_notification_sample(n):
    rc = 0
    global thread_run_flag

    while thread_run_flag == True:
        # 리소스 전송 예제
        rc = example_resource_send_notification();
        if rc < 0 :
            print("FAIL example_resource_send_notification()...");
            thread_run_flag = False
            break;

        print("_thread_notification_sample()...");
        imClient.ImMSleep(1000);


##########################################
# 시작 
##########################################
def iot_sample_run():

    IM_SERVER_ADDR = "dev-gw-iotmakers.cloud.kt.com";
    IM_SERVER_PORT = 30105;
    IM_DEV_ID = "__rbt_ssl_test__";
    IM_DEV_PW = "1234";
    IM_DEV_GW = "ROBOT_SSL_STD";
    IM_LOGLEVEL = 3;    # 1:ERR, 2:INFO, 3:DEBUG


    print("ImInit()...");
    rc = imClient.ImInit(IM_LOGLEVEL);
    if rc < 0 :
        return;

    print("ImSetControlCallBackHandler()...");
    imClient.ImSetControlCallBackHandler(OnResourceSetRequestHandler, OnResourceRetrieveOneRequestHandler, OnResourceRetrieveAllRequestHandler);


    #print("ImSetTlsCaFile()...");
    #imClient.ImSetTlsCaFile("robotmakers.cer.pem");


    print("ImConnectTo()...", IM_SERVER_ADDR, IM_SERVER_PORT);
    rc = imClient.ImConnectTo(IM_SERVER_ADDR, IM_SERVER_PORT);
    if rc < 0 :
        imClient.ImRelease();
        return;


    print("ImAuthDevice()...", IM_DEV_ID, IM_DEV_PW, IM_DEV_GW);
    rc = imClient.ImAuthDevice(IM_DEV_ID, IM_DEV_PW, IM_DEV_GW);
    if rc < 0 :
        imClient.ImDisconnect();
        imClient.ImRelease();
        return;


    #다중쓰레드 고속전송 옵션
    imClient.ImTurnCircuitBreakerOff();
    #imClient.ImTurnResponseWaitOff();

    global thread_run_flag
    thread_run_flag = True

    print("_thread_notification_sample()...");
    t_send = threading.Thread(target=_thread_notification_sample, args=(1,))
    t_send.start()

    print("_thread_poll_sample()...");
    t_poll = threading.Thread(target=_thread_poll_sample, args=(1,))
    t_poll.start()

    while thread_run_flag == True:
        #print("dummy()...");
        imClient.ImMSleep(1000);

    print("_thread_notification_sample()...Join");
    t_send.join()

    print("_thread_poll_sample()...Join");
    t_poll.join()

    print("ImDisconnect()...");
    rc = imClient.ImDisconnect();

    print("ImRelease()...");
    rc = imClient.ImRelease();


if __name__ == '__main__':
    while True:
        iot_sample_run()
        imClient.ImMSleep(1000);

File: project/samples/java/HelloThread.java
/** 
 *  Copyright (c) 2022 KT Corp. All rights reserved.
 *
 *  This is a proprietary software of KT corp, and you may not use this file except in
 *  compliance with license agreement with KT corp. Any redistribution or use of this
 *  software, with or without modification shall be strictly prohibited without prior written
 *  approval of KT corp, and the copyright notice above does not evidence any actual or
 *  intended publication of such software.
 *
 */
import com.kt.iotmakers.device.IoTMakersDeviceClient;

public class HelloThread
{
	///////////////////////////////////////////////////
	// IoTMakers Device init
	///////////////////////////////////////////////////
	IoTMakersDeviceClient cli = new IoTMakersDeviceClient();

	int thread_run_flag = 0;

	/*
	#   리소스 값 전송 예제
	*/
	public int example_resource_send_notification(int hndl)
	{
		String resource_id = "testAllTypeUri";
		String jdata = "{\"data01\":100, \"text01\":\"hello\"}";

		System.out.println(jdata);

		// 1. 전송 패킷 준비
		int rc = cli.ImResourceNotificationInit(hndl);
		if (rc < 0 ) {
			System.out.println("fail ImResourceNotificationInit()");
			return -1;
		}

		// 2. 패킷에 리소스 추가; append FIRST RESOURCE
		rc = cli.ImResourceNotificationAppendResource(hndl, resource_id, jdata);
		if (rc < 0 ) {
			System.out.println("fail ImResourceNotificationAppendResource()");
			return -1;
		}

		// 3. 패킷 전송
		rc = cli.ImResourceNotificationSend(hndl);
		if (rc < 0 ) {
			System.out.println("fail ImResourceNotificationSend()");
			return -1;
		}

		System.out.println("OK, SENT: " + jdata);
		return 0;
	}



	public class ThreadPollSample implements Runnable{
		
		int hndl = 0;

		public ThreadPollSample(int _hndl){
			hndl = _hndl;
		}
		
		@Override public void run() {
			while (thread_run_flag==(1))
			{
				///////////////////////////////////////////////////
				// NetIO polling
				///////////////////////////////////////////////////

				if ( cli.ImPoll(hndl) < 0 ) {
					thread_run_flag = 0;
					System.out.println("FAIL ImPoll()...");
					break;
				}

				cli.ImMSleep(hndl, 1*1000);
			}
			System.out.println("ThreadPollSample()..., DONE.\n");
		}
	}


	public class ThreadNotificationSample implements Runnable{
		
		int hndl = 0;

		public ThreadNotificationSample(int _hndl){
			hndl = _hndl;
		}
		
		@Override public void run() {
			while (thread_run_flag==(1))
			{
				int rc = example_resource_send_notification(hndl);
				if ( rc < 0 ){
					thread_run_flag = 0;
					break;
				}

				cli.ImMSleep(hndl, 1*1000);
			}
			System.out.println("ThreadNotificationSample()..., DONE.\n");
		}
	}




	public void run_main(String  ec_ip, int  ec_port, String dev_id ,String dev_pw, String svc_gw) {
		int ret = 0;

		int log_level = 3;	// 1=ERROR, 2=INFO, 3=DEBUG

		int hndl = cli.ImInit(log_level);
		if ( hndl < 0 )	{
			System.out.println("FAIL cli.ImInit()");
			return;
		}

		///////////////////////////////////////////////////
		// Register Device Control Handler
		///////////////////////////////////////////////////
		cli.setCallBack(hndl, new IoTMakersDeviceClient.ControlCallBack() {
	
            @Override
			public int OnResourceSetRequestHandler(long rep_body, String dev_id, String resource_id, String properties_in_jstr){
				System.out.println("=================================================================");
				System.out.println("OnResourceSetRequestHandler dev_id="+ dev_id +", resource_id="+ resource_id  +", properties_in_jstr="+ properties_in_jstr);
				System.out.println("=================================================================");
				// YOUR CONTROL CODE HERE

				//IM_RESP_CODE_2004_Changed
				return 2004;
            }

            @Override
			public int OnResourceRetrieveOneRequestHandler(long rep_body, String dev_id, String resource_id){
				System.out.println("=================================================================");
				System.out.println("OnResourceRetrieveOneRequestHandler dev_id="+ dev_id +", resource_id="+ resource_id);
				System.out.println("=================================================================");
				
				// YOUR CONTROL CODE HERE
				String jdata = "{\"data01\":100, \"text01\":\"hello\"}";

				int rc = cli.ImResourceRetrieveSetResource(rep_body, resource_id, jdata);
				if (rc < 0 ) {
					System.out.println("fail ImResourceRetrieveSetResource()");
					return 5000;
				}
				
				//IM_RESP_CODE_2000_OK
				return 2000;
            }

            @Override
			public int OnResourceRetrieveAllRequestHandler(long rep_body, String dev_id){
				System.out.println("=================================================================");
				System.out.println("OnResourceRetrieveAllRequestHandler dev_id="+ dev_id);
				System.out.println("=================================================================");
				
				// YOUR CONTROL CODE HERE
				
				String resource_id = "testAllTypeUri";
				String jdata = "{\"data01\":100, \"text01\":\"hello\"}";
				int rc = cli.ImResourceRetrieveAppendResource(rep_body, resource_id, jdata);
				if (rc < 0 ) {
					System.out.println("fail ImResourceRetrieveAppendResource()");
					return 5000;
				}
				
				String resource_id2 = "testAllTypeUri2";
				String jdata2 = "{\"data01\":200, \"text01\":\"world\"}";
				rc = cli.ImResourceRetrieveAppendResource(rep_body, resource_id2, jdata2);
				if (rc < 0 ) {
					System.out.println("fail ImResourceRetrieveAppendResource()");
					return 5000;
				}

				//IM_RESP_CODE_2000_OK
				return 2000;
            }
		
		});


		///////////////////////////////////////////////////
		// -- 다중쓰레드 고속전송 옵션 --
		cli.ImTurnCircuitBreakerOff(hndl); // 1건/초 제한해제
		cli.ImTurnResponseWaitOff(hndl);   // 서버응답 대기안함
		///////////////////////////////////////////////////


		///////////////////////////////////////////////////
		// connect to Server for Open Platform
		//     int ImConnect(int hndl);      
		///////////////////////////////////////////////////
		if ( cli.ImConnectTo(hndl, ec_ip, ec_port) < 0 )	{
			return;
		}

		///////////////////////////////////////////////////
		// Auth Device
		///////////////////////////////////////////////////
		if ( cli.ImAuthDevice(hndl, dev_id, dev_pw, svc_gw) < 0 )	{
			return;
		}


		thread_run_flag = 1;

		Thread t_send = new Thread(new ThreadNotificationSample(hndl));
		t_send.start();
		
		Thread t_poll = new Thread(new ThreadPollSample(hndl));
		t_poll.start();

		while (thread_run_flag==(1))
		{
			cli.ImMSleep(hndl, 1*1000);
		}


		///////////////////////////////////////////////////
		// disconnect
		///////////////////////////////////////////////////
		System.out.println("disconnect hndl = " + hndl);
		cli.ImDisconnect(hndl);

		///////////////////////////////////////////////////
		// release
		///////////////////////////////////////////////////
		System.out.println("release hndl = " + hndl);
		cli.ImRelease(hndl);
	
	}
	



	public static void main(String[] args) {

		/*
		String ec_ip   = "dev-gw-iotmakers.cloud.kt.com";
		int    ec_port = 30105;
		String dev_id  = "testbyw111";
		String dev_pw  = "12312";
		String svc_gw  = "KT_3RD_BYW";
		*/

		String ec_ip   = "dev-gw-iotmakers.cloud.kt.com";
		int    ec_port = 30026;
		String dev_id  = "testbyw111";
		String dev_pw  = "12312";
		String svc_gw  = "KT_3RD_BYW";

		if ( args.length == 5 )
		{
			ec_ip   = args[0];
			ec_port = Integer.parseInt(args[1]);
			dev_id  = args[2];
			dev_pw  = args[3];
			svc_gw  = args[4];
		}

		HelloThread hello = new HelloThread();
		hello.run_main(ec_ip, ec_port, dev_id, dev_pw, svc_gw);
	}


}
